Added token property and filtered the list_ methods

781flyingdutchman opened the pull request:
The token property allows calling client.token and storing that token when the client is created again
in the future (without the need for email/password)

Filtering the list_vehicles and list_energy_sites is required if a user has both a vehicle and an energy
site.  The original implementation throws an error in that case

Dreamsorcerer commented:
Regarding the token property, see my comment here:
https://github.com/mlowijs/tesla_api/issues/16#issuecomment-622955453

This change was only recently merged, but exposing the token directly is likely to result in wrong behaviour when the token expires.

781flyingdutchman commented:
OK that makes sense - thanks for explaining.  I did resubmit (without the token property) because the error in the list_ methods is real and corrected in the latest push.  I also added another API call for the Energy class, and the 'async with' mechanism for easier use of the Client.  I see that commit as a comment above, but not sure if I need to do another pull request for you to see that too - can you let me know? Thanks

Dreamsorcerer commented:
`async with` is a good idea.

See my comments above for some suggested changes.

781flyingdutchman commented:
Sorry, what comments are you referring to?  I removed the token @property per your suggestion, so the changes are now only those described in the commit:
1. Fixed error with list_ methods
2. Added methods to allow async with
3. Add method for calendar history

I think these are good to go, but let me know if not

Dreamsorcerer commented:
You should see them above (nearer the top) in the conversation (plus also inline in the diff).

e.g. `list_vehicles()` shouldn't be changed, and using datetime objects for the new endpoint.

Dreamsorcerer commented:
Oh, my fault, I don't know how to use Github. :P
Seems I have to submit the review...

781flyingdutchman commented on tesla_api/__init__.py:
```diff
@@ -111,7 +121,9 @@ async def post(self, endpoint, data=None):
 
     async def list_vehicles(self, _class=Vehicle):
-        return [_class(self, vehicle) for vehicle in await self.get('vehicles')]
+        return [_class(self, vehicle) for product in await self.get('vehicles') if 'vehicle_id' in product]
```
That may well be true - I couldn't test it (because I have a vehicle and an energy product) so mirrored the logic proposed for the list_energy_sites method

781flyingdutchman commented on tesla_api/__init__.py:
```diff
@@ -111,7 +121,9 @@ async def post(self, endpoint, data=None):
 
     async def list_energy_sites(self, _class=Energy):
-        return [_class(self, products['energy_site_id']) for products in await self.get('products')]
+        return [_class(self, product['energy_site_id']) for product in await self.get('products') if 'energy_site_id' in product]
```
Agree.  I'm also pretty new to GitHub so I don't actually know how I can make this change but not the one for list_vehicles, or should I resubmit the fork altogether?

Dreamsorcerer replied:
I think just revert the change in a new commit, is the usual approach.

781flyingdutchman commented on tesla_api/__init__.py:
```diff
@@ -82,9 +88,11 @@ def _get_headers(self):
         }
 
-    async def get(self, endpoint):
+    async def get(self, endpoint, query=None):
```
I think this is cleaner, and allows for easy extension as there are likely many get methods that include a query of some sort, and  it keeps the meaning of 'endpoint' clean, i.e. it continues to represent the end point, not the tail end of the url

Dreamsorcerer replied:
So far, this is the only one I've seen with query params. But, if we were to support this interface, then I'd expect query_args to be a dictionary, and the full URL to be constructed in this method using that dictionary.

In fact, I just looked at the aiohttp documentation, and we don't even need to do this ourselves. I'd suggest just passing the query params directly to `self._session.get()` below. See: https://docs.aiohttp.org/en/stable/client_quickstart.html#passing-parameters-in-urls

781flyingdutchman replied:
Implemented

781flyingdutchman commented on tesla_api/energy.py:
```diff
@@ -50,10 +55,27 @@ async def get_battery_count(self):
+        Returns historical energy data
+        :param kind: [power, energy, self_consumption]
+        :param period: Amount of time to include in report. One of day, week, month, year, and lifetime.
+                    When kind is power, this parameter is ignored, is not required, and is always treated as day
```
Yes, when `kind` is 'power', the period will always be interpreted as 'day'

Dreamsorcerer commented on tesla_api/energy.py:
```diff
@@ -50,10 +55,27 @@ async def get_battery_count(self):
         return await self._api_client.get('energy_sites/{}/live_status'.format(self._energy_site_id))
 
+    async def get_energy_site_calendar_history_data(self, kind: str = 'energy', period: str = 'day',
+                                                    end_date: str = None, time_zone: str = None):
```
I'm looking to make a new PR soon to add annotations everywhere. But, if you fancy cleaning this up now, then kind/period should use `Literal`.
```
from typing import Literal
...
kind: Literal['power', 'energy', 'self_consumption']
period: Literal['day', 'week', 'month', 'year', 'lifetime']
```

781flyingdutchman replied:
Implemented.  Note weird behavior where if the time is 00:00 the API doesn't return a value, so if you enter a datetime like 2020-05-01 it doesn't work.  As a remedy, if you pass a datetime with the hour at 0 I just add an hour, effectively fetching 2020-05-01T01:00:00 and that works

Dreamsorcerer commented on tesla_api/energy.py:
```diff
@@ -50,10 +59,39 @@ async def get_battery_count(self):
+        """
+        params = {'kind': kind, 'period': period}
+        if end_date is not None:
+            if type(end_date) is datetime:
```
`isinstance(end_date, datetime)` is the right way to do this (correctly handles subclasses etc.).

781flyingdutchman replied:
implemented

Dreamsorcerer commented on tesla_api/energy.py:
```diff
@@ -50,10 +59,39 @@ async def get_battery_count(self):
+                    # so you can use datetime(year=2020, month=5, day=2) and it gets the data for May 2, 2020 as expected
+                    end_date = end_date + timedelta(hours=1)
+                if end_date.tzinfo is None and time_zone is not None:
+                    tz = pytz.timezone(time_zone)
```
Will `datetime.timezone` not work here? pytz is not part of the standard library, so probably don't want to add it as a dependency for this one line.

781flyingdutchman replied:
There are a few issues with datetime.timezone:

1. It does not contain the IANA database for timezone conversions, yet that is the standard used by the Tesla API.  From the python documentation on the datetime.timezone class:
> dateutil.tz library brings the IANA timezone database (also known as the Olson database) to Python, and its usage is recommended
From dateutil is I believe heavier than pytz and from stackoverflow   I understand pytz to be the preferred way to add IANA timezones 
2.  it takes a fixed time offset, whereas pytz correctly applies daylight savings and changes to the timezone over time (e.g., a state deciding to no longer use DST).  Per [this](https://stackoverflow.com/questions/18541051/datetime-and-timezone-conversion-with-pytz-mind-blowing-behaviour) and various stackoverflow articles it is recommended to use 

The alternative is to expect the datetime object to have the timezone set already, but that simply moves the problem back to the user, whereas the IANA timezone is spit out by the Tesla API already when querying the site, so it's easy to add it.  The typical use of the `get_energy_site_calendar_history_data` method becomes:
`response = tesla_client.power_for_day(site_id=site_id, day=datetime(year=2020, month=5, day=1), time_zone='America/Los_Angeles')`

781flyingdutchman replied:
Sorry I added my call to a wrapper around the API.  The call for power for a day becomes:
`response = await site.get_energy_site_calendar_history_data(kind='power', end_date=day, time_zone=time_zone)`
with day simply a datetime like `datetime(year=2020, month=5, day=1)` and the time_zone an IANA string (obtained from the API) such as `time_zone='America/Los_Angeles'`

Dreamsorcerer replied:
I'm failing to find any documentation on these APIs, and can't test it myself. So, just a couple of questions to make sure I understand this correctly.

The timezone actually comes from another endpoint? What code would you write to get that timezone info?

With period set to day, this retrieves the data for a day preceding the given date, but the starting time will always be 01:00 of the preceding day? So, does this go up to the start of the given day, or up to the exact time passed? e.g. If I set the date as 6th 23:59, does that give data for 47 hours (5th 01:00 till 6th 23:59), or does it completely ignore the time and just do till the beginning of the day (5th 01:00 till 6th 01:00)?

781flyingdutchman replied:
I don't think there is formal API documentation (because it is not an officially sanctioned API by Tesla).  The best documentation I have found is [here](https://www.teslaapi.info/energy-sites/state-and-settings#historical-calendar-data), and I have experimented a bit myself.  The timezone data is obtained like this:
`response = await site.get_energy_site_calendar_history_data(kind='power')
            if 'installation_time_zone' in response:
                self.time_zones[site_id] = response['installation_time_zone']`
i.e. calling the api without supplying a date.  Then, you can get the power for one day by calling:
`response = await site.get_energy_site_calendar_history_data(kind='power', end_date=day,
                                                                        time_zone=time_zone)
`
and the reply has a key 'time_series' which returns the power every 15 mins, starting at 0:00.  I don't know why you can't pass a datetime with the hour set to 00:00, but you can't: it doesn't return an error but an empty string - very strange.

781flyingdutchman replied:
Ok, hold on for a minute as I need to check something else with this weird API

Dreamsorcerer replied:
OK, that documentation says that timezone seems to have no effect, so seems there's no point in including it.

So, in my example it would give 47 hours of data in 15 min intervals, right? Weird.

781flyingdutchman replied:
Ok, in the latest push I've deleted the timezone argument altogether.  This means you now have to ensure that the datetime object you pass as end date is timezone aware, otherwise you'll get an error.  I've also changed the increment such that when you pass as end date a datetime without a time (e.g. datetime(year=220, month=5, day=1) you get all data for May 1st as you would expect.  
The raw API call will return data up until the end_date, so if you enter May 1st at 3pm you'll get all data on May 1st from 00:00 to 15:00 in 5 minute intervals.  Normally you would call with a date only, and then get the data for that date (only).  If you call without an end date you get the data for today, from 00:00 to now

Dreamsorcerer commented on tesla_api/energy.py:
```diff
@@ -50,10 +59,39 @@ async def get_battery_count(self):
+                if end_date.tzinfo is None and time_zone is not None:
+                    tz = pytz.timezone(time_zone)
+                    end_date = tz.localize(end_date, time_zone)
+            params['end_date'] = end_date.isoformat() if type(end_date) is datetime else end_date
```
`isinstance()` again.

781flyingdutchman replied:
Implemented

Dreamsorcerer commented on tesla_api/energy.py:
```diff
@@ -50,10 +59,39 @@ async def get_battery_count(self):
+        if end_date is not None:
+            if type(end_date) is datetime:
+                if end_date.hour == 0:
+                    # if the datetime object's time is 00:00 then the API returns nothing. We adjust by adding an hour
```
Hmm, interesting. I don't have a power wall to test this out with, but does 24:00 work, or 00:01?

781flyingdutchman replied:
24:00 generates an API error, anything other than 00:00:00 works.  However, the API response starts at 01:00:00 in the specified timezone (regardless of the time added to the date), so if you're passing a datetime without a time component (as you normally would, given that the API will always return the data for one full day, starting at 1am on that date) it seems appropriate to add one hour if hours==0, as that is most likely the intent.  So, I'm keeping it as is

781flyingdutchman commented:
Last push above - I unfortunately realized that the async/await stuff doesn't play well with Flask, so I cannot use this client for my purpose.  Best of luck!

Dreamsorcerer commented on tesla_api/energy.py:
```diff
@@ -50,10 +56,33 @@ async def get_battery_count(self):
         return await self._api_client.get('energy_sites/{}/live_status'.format(self._energy_site_id))
 
+    async def get_energy_site_calendar_history_data(self, kind: str = 'energy', period: str = 'day',
+                                                    end_date: [str, datetime] = None) -> Optional[dict]:
```
Is there a case where this is going to return `None`? That's what the `Optional` suggests..

Dreamsorcerer replied:
```suggestion
                                                    end_date: Optional[Union[[str, date]] = None) -> Dict[str, Any]:
```

Dreamsorcerer commented on tesla_api/energy.py:
```diff
@@ -50,10 +56,33 @@ async def get_battery_count(self):
+                    Defaults to the current time.
+        """
+        params = {'kind': kind, 'period': period}
+        if end_date is not None:
+            if isinstance(end_date, datetime):
+                if end_date.hour == 0 and end_date.minute == 0:
+                    # if the datetime object's time is 00:00 then the API returns nothing. We adjust by adding 23:59
+                    # so you can use datetime(year=2020, month=5, day=2) and it gets the data for May 2, 2020 as expected
+                    end_date = end_date + timedelta(hours=23, minutes=59)
+            params['end_date'] = end_date.isoformat() if isinstance(end_date, datetime) else end_date
```
Makes sense to support date directly then:
```suggestion
        if isinstance(end_date, date):
            if not isinstance(end_date, datetime):
                end_date = datetime.combine(end_date, time(23, 59, 59))
            elif end_date.hour == 0 and end_date.minute == 0:
                # If the datetime object's time is 00:00 then the API returns nothing. We adjust by adding 23:59,
                # so you can use datetime(year=2020, month=5, day=2) and it gets the data for May 2, 2020 as expected
                end_date = end_date.replace(hour=23, minute=59, second=59)
            end_date = end_date.isoformat()
        params['end_date'] = end_date
```

Dreamsorcerer commented on tesla_api/energy.py:
```diff
@@ -23,13 +23,19 @@
 # SOFTWARE.
 
-import asyncio
+from datetime import datetime, timedelta
+from typing import Optional
```
```suggestion
from datetime import date, datetime, time
from typing import Dict, Optional, Union
```

Dreamsorcerer commented:
You may be able to use it in Flask synchronously by wrapping the calls into `asyncio.get_event_loop().run_until_complete(client.list_vehicles())`, or something along those lines.

Alternatively, if you're in a position to try something a little more modern than Flask, I'd suggest looking at aiohttp, which is a similar interface to Flask but is implemented with Python's asyncio. I've been using it for my own project and it seems to work very well (though I still have to maintain a large Flask project for work).

Dreamsorcerer commented:
Hey, I've got commit access now, so I've merged this. Thanks for the work on this.
